### PR TITLE
Commited in the 48110_user_authentication_with_single_provider

### PR DESCRIPTION
**Pull request by mdmahmudurrahman**

### CODE REVIEW

**Comment by manazirahsan** [comment]
Okey

**Comment by noor-framgia** [comment]
ok

File: config/routes.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Rails.application.routes.draw do
 
-  devise_for :users, controllers: {registrations: "registrations"}
+  devise_for :users, controllers: {registrations: "registrations", omniauth_callbacks:"omniauth_callbacks"}
 
   scope "(:locale)", :locale => /en/ do
     root	'static_pages#home'
```
Comment (shakil-fg) [inline config/routes.rb]: Double quote

File: app/views/devise/shared/_links.html.erb [hunk 2]
```diff
@@ -18,8 +18,3 @@
   <%= link_to "Didn't receive unlock instructions?", new_unlock_path(resource_name) %><br />
```
Comment (shakil-fg) [inline app/views/devise/shared/_links.html.erb]: i18n

File: app/models/user.rb [hunk 3]
```diff
@@ -21,4 +21,38 @@ class User < ActiveRecord::Base
 
   has_many :reports
   enum role: [:supervisor, :trainee]
+
+  def self.from_omniauth(auth)
```
Comment (dieunb) [inline app/models/user.rb]: pls read convention again to write class method

File: app/models/user.rb [hunk 4]
```diff
@@ -21,4 +21,38 @@ class User < ActiveRecord::Base
...
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      # where(auth.slice(:provider, :uid)).first_or_create do |user|
```
Comment (dieunb) [inline app/models/user.rb]: remove all remark line code if you don't use.

File: config/initializers/devise.rb [hunk 5]
```diff
@@ -237,7 +237,7 @@
   # Add a new OmniAuth provider. Check the wiki for more information on setting
   # up on your models and hooks.
   # config.omniauth :github, 'APP_ID', 'APP_SECRET', scope: 'user,public_repo'
-
+    config.omniauth :facebook, "1665605910363611", "975335557c7458c39e01ca6a7593c405"
```
Comment (dieunb) [inline config/initializers/devise.rb]: should store APP_ID and APP_SECRET on environment variable

File: app/views/devise/sessions/new.html.erb [hunk 6]
```diff
@@ -1,26 +1,43 @@
...
   </div>
-
-  <% if devise_mapping.rememberable? -%>
-    <div class="field">
-      <%= f.check_box :remember_me %>
-      <%= f.label :remember_me %>
+  <div class="row">
+    <div class="col-xs-6 col-md-4">
+      <div class="form-group">
+        <%= f.submit "Log in", class:"btn btn-default" %>
```
Comment (dieunb) [inline app/views/devise/sessions/new.html.erb]: i18n

File: app/controllers/omniauth_callbacks_controller.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def all
```
Comment (dieunb) [inline app/controllers/omniauth_callbacks_controller.rb]: do you think about name `provider`?

File: app/models/user.rb [hunk 8]
```diff
@@ -21,4 +21,36 @@ class User < ActiveRecord::Base
 
   has_many :reports
   enum role: [:supervisor, :trainee]
+
+  def self.from_omniauth auth
```
Comment (dieunb) [inline app/models/user.rb]: for class method, should follow
```
class << self
  def from_omniauth
  ...
  end 
end
```

File: app/models/user.rb [hunk 9]
```diff
@@ -21,4 +21,40 @@ class User < ActiveRecord::Base
...
+      where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+        user.provider = auth.provider
+        user.email = auth.info.email
+        user.uid = auth.uid
+        user.name = auth.info.name
+      end
+    end
+  end
+
+  class << self
```
Comment (dieunb) [inline app/models/user.rb]: why you don't merge it with above function?

File: app/models/user.rb [hunk 10]
```diff
@@ -21,4 +21,38 @@ class User < ActiveRecord::Base
 
   has_many :reports
   enum role: [:supervisor, :trainee]
+  
+  class << self
```
Comment (dieunb) [inline app/models/user.rb]: Should
```  
class << self
    def from_omniauth auth
      where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
        user.provider = auth.provider
        user.email = auth.info.email
        user.uid = auth.uid
        user.name = auth.info.name
      end
    end
 
    def new_with_session params, session
      if session["devise.user_attributes"]
        new(session["devise.user_attributes"], without_protection: true) do |user|
          user.attributes = params
          user.valid?
        end
      else
        super
    end
  end
```

**Comment by manazirahsan** [comment]
Okey

**Comment by noor-framgia** [comment]
ok

**Comment by dieunb** [comment]
OK

**Comment by noor-framgia** [comment]
ok

File: app/controllers/omniauth_callbacks_controller.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    user = User.from_omniauth(request.env["omniauth.auth"])
```
Comment (dieunb) [inline app/controllers/omniauth_callbacks_controller.rb]: remove `()`

File: app/models/ability.rb [hunk 12]
```diff
@@ -0,0 +1,14 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
```
Comment (dieunb) [inline app/models/ability.rb]: remove `()`

File: config/application.yml [hunk 13]
```diff
@@ -0,0 +1,3 @@
+FACEBOOK_API_ID: 1665605910363611
```
Comment (dieunb) [inline config/application.yml]: don't push this file to git

**Comment by manazirahsan** [comment]
ok

**Pull request closed without merging**